### ISSUE TITLE
*azservicebus.Receiver.ReceiveMessages(ctx, 1, nil) might not return messages on success

### ISSUE THREAD

**Issue by henriksjostrom** [Azure/azure-sdk-for-go#21017]
### Bug Report

- import path of package in question: github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus
- version: v1.4.0
from go.sum: github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus v1.4.0 h1:MxbPJrYY81a8xnMml4qICSq1z2WusPw3jSfdIMupnYM=
- output of `go version` go version go1.20 linux/amd64

**- What happened?** 
```golang
func (rec *azservicebus.Receiver, ctx context.Context) (*azservicebus.ReceivedMessage, error){
	messages, err := amq.receiver.ReceiveMessages(ctx, 1, nil)
	if err != nil {
		return nil, err
	}
	if len(messages) != 1 {
		panic("received %d messages", len(messages))
	}
	return messages[0], nil
}
```
can panic because len(messages)=0 when there is an internal error in the servicebus. Errors related to service request #2306020050002066.

At the same time in Typescript the following error is returned which is far more accurate than not returning an error at all.
```

ServiceBusError: InternalServerError: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 TrackingId:<removed>, SystemTracker:gateway5, Timestamp:2023-06-13T10:31:44
    
at translateServiceBusError (/opt/app/node_modules/@azure/service-bus/dist/index.js:4387:16)
...
```

**- What did you expect or want to happen?**
I'd expect this function to return an error describing the problem, preferably something more in line with what TS does.

**- How can we reproduce it?**
Setup the library to listen, 
cause InternalServerError in service bus
see if error is returned or not.

**- Anything we should know about your environment.**
No

**Comment by henriksjostrom** [comment]
@richardpark-msft 

I'm unaware of what causes the internal error in the service bus. Its being investigated as errand #2306020050002066 by the product team. 

I'm pretty sure the context isn't timing out as we aren't setting it to time out. 

How could we see if the link is detached? This could be whats happening is the server is being crashy. If there is no error returned how do we check for that?

**Comment by richardpark-msft** [comment]
> @richardpark-msft Richard Park (DEVDIV) FTE

> 
> I'm pretty sure the context isn't timing out as we aren't setting it to time out.
> 
> How could we see if the link is detached? This could be whats happening is the server is being crashy. If there is no error returned how do we check for that?
> 

You can turn on our debug logging ([link](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/messaging/azservicebus#logging)), which will let you get full visibility into what failures are happening, even if we actually did recover. It's a bit verbose, but it does give a lot of information.

> This is an error on the TypeScript side (or at least the error happens at the same time every time)

It's possible our error handling policies are different based on our different programming models, although I can't be sure. I think the logging will give us what we need.

How long does this typically take to replicate?

**Comment by henriksjostrom** [comment]
it happens about once per week or so

**Comment by richardpark-msft** [comment]
BTW, if you remove your panic it's safe to continue to receive - the receiver will just recover on its own. This might also help us to see if the condition itself is fatal, or temporary.

**Comment by henriksjostrom** [comment]
Apparently this happens whenever there is a new queue deployment by MS and the connection dies. 

Should it return no messages or should it return some kind of informative error?

**Comment by henriksjostrom** [comment]
When logging the events described on that page we get the following right before we get the 0 message issue

chronological order first message first and then descending
```
[c:1, l:2, r:name:hFPSh5] Received 0/1 messages
[c:1, l:2, r:name:hFPSh5] Closing links for error amqp: link closed
[c:1, l:2, r:name:hFPSh5] Links closing (permanent: false)
[c:1, l:2, r:name:hFPSh5] Closing Receiver
[c:1, l:2, r:name:hFPSh5] Closing Session
[c:1, l:2, r:name:hFPSh5] Closing RPC
[c:1, l:2, r:name:hFPSh5] Failure when receiving messages: amqp: link closed
```

**Comment by richardpark-msft** [comment]
> Apparently this happens whenever there is a new queue deployment by MS and the connection dies.
> 
> Should it return no messages or should it return some kind of informative error?

TLDR: The best way to use the library is in a loop. Zero messages, no error just means that you should continue to try to receive.

I'm going to close this now that we understand the cause and that the behavior is expected, but please reopen if you have more questions.

Longer:

The general idea is we don't want to return an error unless you are expected to do something about it - cosmetic/informational errors generally just mean extra code to handle a case where all we want you to do is retry.

This particular scenario is one of those cases - the link has detached, which is fine, so we'll reattach it at the next call. 

There are some scenarios that return errors but those are in cases where the error is "fatal" (for instance, you asked us to complete a message but the lock has expired), or if the connection can't be recovered within your configured number of retries.
